### DISCUSSION TITLE
另一个谷歌 Play商店下载问题的 DNS 配置

### DISCUSSION THREAD

**Discussion by mustang5910** [kkkgo/PaoPaoDNS#139] [经验分享]
不知道什么原因,  我在使用 PaoPaoDNS + PaoPaoGateway 时, google play 商店能浏览但无法下载更新应用.
仔细研读说明和配置后, 发现 PaoPaoDNS 的默认配置转发规则, 在我的环境上是不可用的. 
最后摸索出下面的配置片段后, 总归是能正常使用Play商店了. 故在此分享给和我有一样困扰的大家

PS: 也许和我访问 google 的 IP 被 google 标记为 CN 有关联

force_forward_list.txt
``` 
# Google play download issues
domain:google.cn
domain:googleapis.cn
# domain:xn--ngstr-lra8j.com
domain:gvt1.com
domain:android.googleapis.com
domain:play.googleapis.com
```
force_dnscrypt_list.txt
```
domain:gvt1-cn.com
domain:googlevideo.com
domain:xn--ngstr-lra8j.com
full:connectivitycheck.gstatic.com
```

**Comment by kkkgo** [comment]
如果你开启了`USE_MARK_DATA`功能，应该包含了这些域名。

**Reply by kkkgo** [reply]
另外force_forward_list优先级>force_dnscrypt_list>AUTO_FORWARD，按你的force_dnscrypt_list来看应该无法访问googlevideo了。

**Reply by mustang5910** [reply]
有意思的地方就是这里, force_dnscrypt_list.txt 这的 4个域(包括googlevideo) , 我这里能直连访问且下载速度飞快, 但是如果交给代理访问后,   反而会断联....

**Reply by mustang5910** [reply]
所以我不认为这会是个通用的配置,  应该是某种特定环境干扰造成的特定解. 甚至可能只对我有效.......

**Reply by mustang5910** [reply]
补一个 docker-compose.yml
``` docker-compose.yml
version: "3"

services:
  paopaodns:
    image: sliamb/paopaodns:latest
    restart: always
    volumes:
      - ./paopaodns:/data
    environment:
      - TZ=Asia/Shanghai
      - UPDATE=weekly
      - DNS_SERVERNAME=PaoPaoDNS
      - DNSPORT=53
      - CNAUTO=yes
      - CNFALL=no
      - CN_TRACKER=yes
      - USE_HOSTS=no
      - IPV6=yes
      - SOCKS5=yyy.yyy.yyy.yyy:1080
      - SERVER_IP=xxx.xxx.xxx.xxx
      - CUSTOM_FORWARD=yyy.yyy.yyy.yyy:53
      - CUSTOM_FORWARD_TTL=300
      - AUTO_FORWARD=yes
      - AUTO_FORWARD_CHECK=yes
      - USE_MARK_DATA=yes
      - RULES_TTL=300
      - ADDINFO=yes
      - SHUFFLE=trnc
      - HTTP_FILE=no
    network_mode: host
    privileged: true
```

**Comment by mustang5910** [comment]
![屏幕截图 2024-05-07 144245](https://github.com/kkkgo/PaoPaoDNS/assets/31264658/23d30eae-3847-4545-8bcc-e964cf927982)
贴个图证明我确实能直连访问 googlevideo.com 的子域

**Reply by kkkgo** [reply]
这是谷歌中国特供的开发工具下载镜像，节点就在北京上海的，那肯定能访问。
你去访问youtube走的googlevideo就不能了。

**Comment by mouyase** [comment]
我也遇到了类似的问题，PlayStore里会出现一直在等待中的情况，也许并不是个例

**Reply by kkkgo** [reply]
考虑排查几种情况：
1、被分配到IPv6地址或者DNS，被谷歌优先使用
2、play商店使用了谷歌DNS，不走DNS解析
3、谷歌商店部分内容使用了UDP协议（QUIC），代理节点不支持UDP或者连接性不佳。
4、可能有没有被成功代理的域名/IP段

情况1和2可以参考https://github.com/kkkgo/PaoPaoGateWay/discussions/98#discussioncomment-9485219
情况3可以设置ppgw禁用udp：`udp_enable=no`，或者使用docker定制`sniff=yes`的镜像(sniff会禁用QUIC)
情况4可以把手机设置静态IP，设置DNS和网关地址都为ppgw，这个情况下是全局代理，如果正常使用，观察日志或者流量记录通过的域名/IP段。

**Reply by mustang5910** [reply]
虽然没有证据， 但是我现在更加怀疑是 代理 IP 被 google 标记为 CN 的原因， 逻辑应该大致是这样的， 
```
代理请求 google cdn --->  google 分发 CN cdn 给代理 ---> 代理无法访问 CN 的 cdn 节点 ---> google play 等待中无法下载。 
```
主要根源应该是 play 服务的位置服务长期向 google 汇报位置在中国，导致代理 IP  被 google 标记为 CN 
也许使用机场的朋友应该不会遇到这个情况
我自己的解决方法比较粗暴， 就是将 domain:googlevideo.com 加入直连列表， 代价是如 kkkgo 所说不能上 youtube 看视频

**Reply by mustang5910** [reply]
更精细的做法可以把google CN 的cdn单独分开加进直连列表试试， 比如正则 regexp:.\*sn-ni57rn7.\*.googlevideo.com

**Reply by kkkgo** [reply]
自建节点的话，你可以在rules:部分添加规则：GEOIP,CN,DIRECT

**Reply by kkkgo** [reply]
[surfboard](https://github.com/getsurfboard/surfboard/releases)似乎自带了一个检测工具，你可以验证下。
![image](https://github.com/kkkgo/PaoPaoDNS/assets/4530897/024fa534-85bf-4092-9a53-257ac3d5598e)

**Comment by Edmondguo** [comment]
截至目前，这个问题解决了吗？

**Reply by woopy-ming** [reply]
可以去看看openclash作者有解释过这事情，主要是因为services.googleapis.cn 等 Google 域名被国内 DNS 解析到中国大陆 IP（220.181.x.x），被 china_ip_route 规则匹配后走直连；但 Google 中国服务器禁止境外 IP（代理节点）访问，导致死循环，这是他的解释，把这些地址改为用海外dns解析就行

**Reply by Edmondguo** [reply]
我用 dig services.googleapis.cn 返回的是fake-ip，是不是说明不是走得直连

**Reply by woopy-ming** [reply]
应该是了，因为fakeip是你的forward服务器返回的，看你那边是怎么设置的，看看那边的日志就知道了。别让直连就行，这网址只要直连哪怕用8888dns返回的都是国内ip
